### DISCUSSION TITLE
ERROR ~ * Missing required parameter

### DISCUSSION THREAD

**Discussion by kentaro0209** [PGScatalog/pgsc_calc#204] [Q&A]
Hello,
I'm a beginner of coding. I need your help to resolve the issue in running pgsc_calc as below.

kentaro0209@DESKTOP-ATMBSV9:~$ nextflow run pgscatalog/pgsc_calc \
    -profile docker
    --input /home/kentaro0209/samplesheet.csv --target_build GRCh38 \
    --pgs_id PGS002814 \
N E X T F L O W  ~  version 23.10.0
Launching `https://github.com/pgscatalog/pgsc_calc` [astonishing_wing] DSL2 - revision: 863166865e [main]

ERROR ~ ERROR: Validation of pipeline parameters failed!

 -- Check '.nextflow.log' file for details
ERROR ~ * Missing required parameter: --input

 -- Check '.nextflow.log' file for details
ERROR ~ * Missing required parameter: --target_build

 -- Check '.nextflow.log' file for details

I'm guessing the directory of the samplesheet file might not be correct. Where should I put the file?
The profile test was successfully done.

Thank you for your help!

Ken

**Comment by nebfield** [comment]
I think removing the newlines might help:

```
$ nextflow run pgscatalog/pgsc_calc -profile docker --input /home/kentaro0209/samplesheet.csv --target_build GRCh38 --pgs_id PGS002814 
```

(if you're copying and pasting into your terminal, don't include the `$`)

**Comment by kentaro0209** [comment]
Thanks! It works!
Ken

**Comment by kentaro0209** [comment]
Hello,

I encountered another error in the following process as below. I ran the same code.

!! Only displaying parameters that differ from the pipeline defaults !!
------------------------------------------------------
If you use pgscatalog/pgsc_calc for your analysis please cite:

* The Polygenic Score Catalog
  https://doi.org/10.1038/s41588-021-00783-5

* The nf-core framework
  https://doi.org/10.1038/s41587-020-0439-x

* Software dependencies
  https://github.com/pgscatalog/pgsc_calc/blob/master/CITATIONS.md
------------------------------------------------------
executor >  local (6)
executor >  local (6)
executor >  local (7)
[5a/3f2d5d] process > PGSCATALOG_PGSCALC:PGSCALC:DOWNLOAD_SCOREFILES ([... [100%] 1 of 1 ✔
executor >  local (7)
[5a/3f2d5d] process > PGSCATALOG_PGSCALC:PGSCALC:DOWNLOAD_SCOREFILES ([... [100%] 1 of 1 ✔
executor >  local (7)
[5a/3f2d5d] process > PGSCATALOG_PGSCALC:PGSCALC:DOWNLOAD_SCOREFILES ([... [100%] 1 of 1 ✔
[6a/d03e7d] process > PGSCATALOG_PGSCALC:PGSCALC:INPUT_CHECK:SAMPLESHEE... [100%] 1 of 1 ✔
[1b/50510f] process > PGSCATALOG_PGSCALC:PGSCALC:INPUT_CHECK:COMBINE_SC... [100%] 1 of 1 ✔
[skipped  ] process > PGSCATALOG_PGSCALC:PGSCALC:MAKE_COMPATIBLE:PLINK2... [100%] 1 of 1, stored: 1 ✔
[-        ] process > PGSCATALOG_PGSCALC:PGSCALC:MAKE_COMPATIBLE:PLINK2... -
[-        ] process > PGSCATALOG_PGSCALC:PGSCALC:MAKE_COMPATIBLE:PLINK2... -
[45/ec0fe5] process > PGSCATALOG_PGSCALC:PGSCALC:MATCH:MATCH_VARIANTS (... [100%] 1 of 1 ✔
[32/a9adb3] process > PGSCATALOG_PGSCALC:PGSCALC:MATCH:MATCH_COMBINE (EUR) [ 66%] 2 of 3, failed: 2..
[-        ] process > PGSCATALOG_PGSCALC:PGSCALC:APPLY_SCORE:PLINK2_SCORE  -
[-        ] process > PGSCATALOG_PGSCALC:PGSCALC:APPLY_SCORE:SCORE_AGGR... -
[-        ] process > PGSCATALOG_PGSCALC:PGSCALC:REPORT:SCORE_REPORT       -
[-        ] process > PGSCATALOG_PGSCALC:PGSCALC:DUMPSOFTWAREVERSIONS      -
[skipping] Stored process > PGSCATALOG_PGSCALC:PGSCALC:MAKE_COMPATIBLE:PLINK2_RELABELBIM (1)
[3a/da045f] NOTE: Process `PGSCATALOG_PGSCALC:PGSCALC:MATCH:MATCH_COMBINE (EUR)` terminated with an error exit status (1) -- Execution is retried (1)
[e9/93b659] NOTE: Process `PGSCATALOG_PGSCALC:PGSCALC:MATCH:MATCH_COMBINE (EUR)` terminated with an error exit status (1) -- Execution is retried (2)
ERROR ~ Error executing process > 'PGSCATALOG_PGSCALC:PGSCALC:MATCH:MATCH_COMBINE (EUR)'

Caused by:
  Process `PGSCATALOG_PGSCALC:PGSCALC:MATCH:MATCH_COMBINE (EUR)` terminated with an error exit status (1)

Command executed:

  export POLARS_MAX_THREADS=2

  combine_matches                  --dataset EUR         --scorefile scorefiles.txt.gz         --matches *.ipc.zst         -n 2         --min_overlap 0.75                                    --outdir $PWD                  --combined         -v

  cat <<-END_VERSIONS > versions.yml
  MATCH_COMBINE:
      pgscatalog_utils: $(echo $(python -c 'import pgscatalog_utils; print(pgscatalog_utils.__version__)'))
  END_VERSIONS

Command exit status:
  1

Command output:
  (empty)

Command error:
  root: 2023-10-27 06:32:06 DEBUG    Verbose logging enabled
  
  I really appreciate your help and support!
  
  Ken

**Reply by kentaro0209** [reply]
Additional information from nextflow.log

Caused by:
  Process `PGSCATALOG_PGSCALC:PGSCALC:MATCH:MATCH_COMBINE (EUR)` terminated with an error exit status (1)

Command executed:

  export POLARS_MAX_THREADS=2
  
  combine_matches                  --dataset EUR         --scorefile scorefiles.txt.gz         --matches *.ipc.zst         -n 2         --min_overlap 0.75                                    --outdir $PWD                  --combined         -v
  
  cat <<-END_VERSIONS > versions.yml
  MATCH_COMBINE:
      pgscatalog_utils: $(echo $(python -c 'import pgscatalog_utils; print(pgscatalog_utils.__version__)'))
  END_VERSIONS

Command exit status:
  1

Command output:
  (empty)

Command error:
  root: 2023-10-27 06:32:06 DEBUG    Verbose logging enabled
  pgscatalog_utils.config: 2023-10-27 06:32:06 DEBUG    Using 2 threads to read CSVs
  pgscatalog_utils.config: 2023-10-27 06:32:06 DEBUG    polars threadpool size: 2
  pgscatalog_utils.match.read: 2023-10-27 06:32:06 DEBUG    Reading scorefile
  pgscatalog_utils.match.read: 2023-10-27 06:32:06 DEBUG    --chrom parameter not set, using all variants in scoring file
  pgscatalog_utils.match.preprocess: 2023-10-27 06:32:06 DEBUG    Complementing column effect_allele
  pgscatalog_utils.match.preprocess: 2023-10-27 06:32:06 DEBUG    Complementing column other_allele
  pgscatalog_utils.match.combine_matches: 2023-10-27 06:32:06 DEBUG    Reading matches
  pgscatalog_utils.match.combine_matches: 2023-10-27 06:32:06 DEBUG    Labelling match candidates
  pgscatalog_utils.match.label: 2023-10-27 06:32:06 DEBUG    Labelling best match type (refalt > altref > ...)
  pgscatalog_utils.match.label: 2023-10-27 06:32:06 DEBUG    Labelling duplicated best match: keeping first instance as best_match = True
  pgscatalog_utils.match.label: 2023-10-27 06:32:06 DEBUG    Labelling multiple scoring file lines (accession/row_nr) that best_match to the same variant
  pgscatalog_utils.match.label: 2023-10-27 06:32:06 DEBUG    Labelling all duplicates with exclude flag
  pgscatalog_utils.match.label: 2023-10-27 06:32:06 DEBUG    Labelling ambiguous variants
  pgscatalog_utils.match.preprocess: 2023-10-27 06:32:06 DEBUG    Complementing column REF
  pgscatalog_utils.match.label: 2023-10-27 06:32:06 DEBUG    Labelling ambiguous variants with exclude flag
  pgscatalog_utils.match.label: 2023-10-27 06:32:06 DEBUG    Labelling multiallelic matches with exclude flag
  pgscatalog_utils.match.label: 2023-10-27 06:32:06 DEBUG    Not excluding flipped matches
  pgscatalog_utils.match.filter: 2023-10-27 06:32:06 DEBUG    Filtering to best_match variants (with exclude flag = False)
  pgscatalog_utils.match.filter: 2023-10-27 06:32:06 DEBUG    Calculating overlap between target genome and scoring file
  pgscatalog_utils.match.filter: 2023-10-27 06:32:06 ERROR    Score PGS002814_hmPOS_GRCh38 fails minimum matching threshold (0.04% variants match)
  pgscatalog_utils.match.match_variants: 2023-10-27 06:32:06 CRITICAL Error: no target variants match any variants in scoring files
  Traceback (most recent call last):
    File "/venv/bin/combine_matches", line 8, in <module>
      sys.exit(combine_matches())
    File "/venv/lib/python3.10/site-packages/pgscatalog_utils/match/combine_matches.py", line 40, in combine_matches
      log_and_write(matches=matches, scorefile=scorefile, dataset=dataset, args=args)
    File "/venv/lib/python3.10/site-packages/pgscatalog_utils/match/match_variants.py", line 90, in log_and_write
      raise Exception("No valid matches found")
  Exception: No valid matches found

**Reply by nebfield** [reply]
I'm glad it's working now. You're experiencing a common error where the genetic variants in your target genomes aren't present in the scoring files. See https://github.com/PGScatalog/pgsc_calc/issues/86#issuecomment-1433289298 for a longer explanation

This is a problem caused by your input data. The most common causes are you accidentally set the wrong genome build or you forgot to impute your genomes.

**Reply by kentaro0209** [reply]
Thanks, 
I'm trying another bfile datasets as input. Imputation was done for this datasets.
Another error happens as below. 

  root: 2023-11-01 01:31:41 DEBUG    Verbose logging enabled
  pgscatalog_utils.config: 2023-11-01 01:31:41 DEBUG    Using 2 threads to read CSVs
  pgscatalog_utils.config: 2023-11-01 01:31:41 DEBUG    polars threadpool size: 2
  pgscatalog_utils.match.read: 2023-11-01 01:31:41 DEBUG    Reading scorefile
  pgscatalog_utils.match.read: 2023-11-01 01:31:41 DEBUG    --chrom parameter not set, using all variants in scoring file
  pgscatalog_utils.match.preprocess: 2023-11-01 01:31:41 DEBUG    Complementing column effect_allele
  pgscatalog_utils.match.preprocess: 2023-11-01 01:31:41 DEBUG    Complementing column other_allele
  pgscatalog_utils.match.combine_matches: 2023-11-01 01:31:41 DEBUG    Reading matches
  pgscatalog_utils.match.combine_matches: 2023-11-01 01:31:41 DEBUG    Labelling match candidates
  pgscatalog_utils.match.label: 2023-11-01 01:31:41 DEBUG    Labelling best match type (refalt > altref > ...)
  pgscatalog_utils.match.label: 2023-11-01 01:31:41 DEBUG    Labelling duplicated best match: keeping first instance as best_match = True
  pgscatalog_utils.match.label: 2023-11-01 01:31:41 DEBUG    Labelling multiple scoring file lines (accession/row_nr) that best_match to the same variant
  pgscatalog_utils.match.label: 2023-11-01 01:31:41 DEBUG    Labelling all duplicates with exclude flag
  pgscatalog_utils.match.label: 2023-11-01 01:31:41 DEBUG    Labelling ambiguous variants
  pgscatalog_utils.match.preprocess: 2023-11-01 01:31:41 DEBUG    Complementing column REF
  pgscatalog_utils.match.label: 2023-11-01 01:31:41 DEBUG    Labelling ambiguous variants with exclude flag
  pgscatalog_utils.match.label: 2023-11-01 01:31:41 DEBUG    Labelling multiallelic matches with exclude flag
  pgscatalog_utils.match.label: 2023-11-01 01:31:41 DEBUG    Not excluding flipped matches
  Traceback (most recent call last):
    File "/venv/bin/combine_matches", line 8, in <module>
      sys.exit(combine_matches())
    File "/venv/lib/python3.10/site-packages/pgscatalog_utils/match/combine_matches.py", line 37, in combine_matches
      _check_duplicate_vars(matches)
    File "/venv/lib/python3.10/site-packages/pgscatalog_utils/match/combine_matches.py", line 52, in _check_duplicate_vars
      assert max_occurrence == [1], "Duplicate IDs in final matches"
  AssertionError: Duplicate IDs in final matches
  
  What "duplicate IDs in final matches"? I checked the discussion #199 (https://github.com/PGScatalog/pgsc_calc/discussions/199), and try Grch37/38 as --target_build. But both didn't work.
  
Thank you for your help and support!

**Reply by nebfield** [reply]
If your data have one file per chromosome, you might want to double check your samplesheet.

The chrom column of your samplesheet shouldn't contain the `chr` prefix (`chr1` ❌, `1` ✅). If you have the chr prefix then this unhelpful error can happen 😅 

There is also a chance that you need to manually deduplicate your genotypes with [`plink2 --rm-dup`](https://www.cog-genomics.org/plink/2.0/filter) before you can use the calculator.

**Reply by kentaro0209** [reply]
mmm... My samplesheet doesn't contain chr prefix.
And I tried plink2 --rm-dup, but didn't work.

**Reply by nebfield** [reply]
If you're:

* 100% sure the genome build matches the parameter `--target_build`
* [processed your target genomes following our recommendations](https://pgsc-calc.readthedocs.io/en/latest/how-to/prepare.html)
* imputed your target genomes 

and you're still experiencing problems, I'm not sure what else to suggest, sorry 😢 You could possibly test a combined file (i.e. all chromosomes in one file) to see if that helps generating a more informative error message.

We're going to do a patch in a few weeks to improve the error messages which may help track down the problem

**Reply by kentaro0209** [reply]
Thank you for your comments. We figured out there was a problem on our bim file. The 4th column was somehow all 0 value. We entered corresponding values to 4th column, and it worked. Thank you for your help!
